### PR TITLE
feat: migrate CampaignEditor test to React Testing Library

### LINKED ISSUE #343: Refactor: migrate CampaignEditor.test.tsx to RTL with query and interaction quality improvements
## Context

Part of the page-level RTL migration (#263). `CampaignEditor.test.tsx` is the largest and most complex of the four files (387 lines, 25+ `act()` calls). Beyond the mechanical pattern swap, this file has several fragile query patterns and low-level interaction hacks that should be fixed during migration — not just carried forward in RTL style.

**Prerequisite:** #263 (RTL infrastructure, `@testing-library/react` and `@testing-library/user-event` installed)

---

### Boilerplate
- `@jest-environment jsdom` docblock (covered by #264, but remove here too)
- `(globalThis as unknown as ...).IS_REACT_ACT_ENVIRONMENT = true` — RTL sets this automatically
- `createReactRoot` / `unmountReactRoot` imports and usage — replaced by RTL's `render` + automatic cleanup
- Local `render()` wrapper around `act(() => root.render(...))` — replaced by RTL `render()`
- `findButton(text)` helper — replaced by `screen.getByRole('button', { name: /text/ })`
- `getInput(type, index)` helper — defined but unused; delete it
- `openChapters()` helper — replace with an inline RTL-based version or inline per-test

### Imports
Remove: `React`, `Root`, `act`, `createReactRoot`, `unmountReactRoot`
Add: `render`, `screen`, `cleanup` from `@testing-library/react`; `userEvent` from `@testing-library/user-event`

---

### 1. Positional text input queries (lines 85, 90)
```ts
// Before — breaks silently if any text input is added above or reordered
container.querySelectorAll<HTMLInputElement>('input[type="text"]')[0].value
container.querySelectorAll<HTMLInputElement>('input[type="text"]')[1].value

// After — query by label or accessible name
screen.getByRole('textbox', { name: /campaign name/i })
screen.getByRole('textbox', { name: /module/i })
// Or: screen.getByDisplayValue('Test Campaign')
```
If the inputs lack accessible labels, add them to the component as part of this migration.

### 2. Broad `container.textContent` assertions (lines 113, 204–206, 228–231, 240, 248)
```ts
// Before — matches anywhere in the DOM, including hidden text
expect(container.textContent).toContain('Arrival');

// After — assert presence in the rendered output specifically
expect(screen.getByText('Arrival')).toBeInTheDocument();
// Or for negative: expect(screen.queryByText('+ Add Chapter')).not.toBeInTheDocument();
```

### 3. Label queries via raw DOM iteration (lines 182–190)
```ts
// Before — brittle manual array map
const labels = Array.from(container.querySelectorAll('label')).map(l => l.textContent);
expect(labels.some(l => l?.includes('Current Chapter'))).toBe(false);

// After
expect(screen.queryByText(/current chapter/i)).not.toBeInTheDocument();
```

### 4. Numeric index in `data-testid` (lines 265, 293, 301, 376)
```ts
// Before — test ID encodes position; breaks on reorder
container.querySelector('button[data-testid="remove-chapter-1"]')
container.querySelector('button[data-testid="move-up-1"]')
container.querySelector('button[data-testid="move-down-0"]')

// After — query by accessible name tied to chapter title
screen.getByRole('button', { name: /remove the inn/i })
screen.getByRole('button', { name: /move the inn up/i })
```
This requires the component to set `aria-label` on these buttons (e.g., `aria-label="Remove The Inn"`). Add those labels to the component as part of this migration — it also improves screen reader accessibility.

---

### 5. Manual `select.value` + `dispatchEvent` (lines 149–152, 161–164, 332–335)
```ts
// Before — bypasses React's synthetic event system; may not trigger React state updates correctly
select.value = 'completed';
select.dispatchEvent(new Event('change', { bubbles: true }));

// After — fires real browser-like events through userEvent
await userEvent.selectOptions(screen.getByRole('combobox', { name: /status/i }), 'completed');
```

### 6. Manual `input.value` + `dispatchEvent` (lines 357–360)
```ts
// Before — same problem: mutates DOM directly, bypassing React
input.value = 'New Arrival';
input.dispatchEvent(new Event('change', { bubbles: true }));

// After
const input = screen.getByRole('textbox', { name: /chapter title/i });
await userEvent.clear(input);
await userEvent.type(input, 'New Arrival');
```

### 7. All `await act(async () => { btn.click() })` patterns (~20 occurrences)
```ts
// Before
await act(async () => { findButton('Save Campaign').click(); });

// After
await userEvent.click(screen.getByRole('button', { name: /save campaign/i }));
```

---

## `openChapters()` Helper

The current helper checks `container.textContent?.includes('+ Add Chapter')` to decide whether to click the accordion. In RTL:

```ts
async function openChapters() {
  if (!screen.queryByText('+ Add Chapter')) {
    await userEvent.click(screen.getByRole('button', { name: /chapters/i }));
  }
}
```

Consider whether this helper is worth keeping or whether each test that needs chapters open should just call `userEvent.click` directly — the accordion is a single click, not a complex setup.

---

## Component Changes Required

This migration surfaces two component accessibility gaps that need to be fixed:

1. **Name and Module Name inputs** — add `<label>` elements (or `aria-label`) so they can be queried by accessible name
2. **Chapter action buttons** (remove, move up, move down) — add `aria-label` values that include the chapter title, e.g. `aria-label={`Remove ${chapter.title}`}`

These are small changes but they make the component more accessible and the tests more robust.

---

## Suggested Test Structure After Migration

```ts
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { CampaignEditor } from '@/app/campaigns/CampaignEditor';

// No beforeEach/afterEach needed — RTL cleanup is automatic

function renderEditor(overrides = {}) {
  const props = {
    campaign: { ...BASE_CAMPAIGN, ...overrides },
    onSave: jest.fn(),
    onCancel: jest.fn(),
    isNew: false,
  };
  render(<CampaignEditor {...props} />);
  return props; // return mocks for assertion
}
```

---

## Acceptance Criteria

- [ ] All `createRoot`, `act`, `IS_REACT_ACT_ENVIRONMENT`, `createReactRoot`, `unmountReactRoot` removed
- [ ] `@testing-library/react` and `@testing-library/user-event` used exclusively for render and interaction
- [ ] No positional `querySelectorAll(...)[index]` queries
- [ ] No `container.textContent.includes(...)` broad assertions
- [ ] No `element.value = ...; dispatchEvent(...)` interaction hacks — all replaced with `userEvent`
- [ ] Chapter action buttons have `aria-label` values in the component
- [ ] Name/Module Name inputs have accessible labels in the component
- [ ] All 25 existing test cases pass
- [ ] CI passes

### PR DESCRIPTION
**Pull request by dougis**
## Summary

Migrates `tests/unit/components/CampaignEditor.test.tsx` from manual DOM/`act`/`createReactRoot` patterns to `@testing-library/react` + `@testing-library/user-event` v14.

Also adds two small accessibility improvements to source components required to make RTL role-based queries work cleanly.

### `lib/components/ui.tsx`
- Added `slugify` helper that converts a label string to a URL-safe slug
- `TextInputField` now auto-generates an accessible `id` from its `label` prop when no explicit `id` is provided, with a `useId()` fallback for edge cases
- This enables `screen.getByRole('textbox', { name: /campaign name/i })` queries in tests

### `app/campaigns/CampaignEditor.tsx`
- Chapter title `<input>` elements now have `aria-label="Chapter N title"` 
- Chapter remove `<button>` elements now have `aria-label="Remove <title|chapter N>"`
- Move-up/move-down buttons already had correct `aria-label` values (verified)

### `tests/unit/components/CampaignEditor.test.tsx`
- Removed: `@jest-environment jsdom` docblock, `IS_REACT_ACT_ENVIRONMENT`, `createReactRoot`/`unmountReactRoot`, `React`, `act` imports
- Removed: `beforeEach` container/root setup, `afterEach` unmount, `findButton()`, `getInput()`, manual `render()` wrapper
- Added: RTL `render`, `screen`; `userEvent.setup()` v14 pattern
- Added: `renderEditor()` factory returning `{ onSave, onCancel, user }`
- All 26 tests migrated to semantic role/label queries; zero `.click()`, `.value =`, or `dispatchEvent` patterns remain

## Validation

- `npx jest --testPathPattern CampaignEditor` → **26/26 pass**
- `npx jest --testPathPattern tests/unit` → **1901/1901 pass** (all unit suites)
- `npx tsc --noEmit` → clean
- `npm run build` → success

Closes #343

### CODE REVIEW

File: openspec/changes/migrate-campaigneditor-test-to-rtl/tasks.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
+# Tasks
+
+## Preparation
+
+- [x] **Step 1 — Sync default branch:** `git checkout main` and `git pull --ff-only`
+- [x] **Step 2 — Create and publish working branch:** `git checkout -b feat/migrate-campaigneditor-test-to-rtl` then immediately `git push -u origin feat/migrate-campaigneditor-test-to-rtl`
```
[the remark this answers is not in the document]
Comment (dougis) [reply openspec/changes/migrate-campaigneditor-test-to-rtl/tasks.md:6]: This is working as designed. The `openspec/changes/migrate-campaigneditor-test-to-rtl/` directory contains all expected artifacts (`.openspec.yaml`, `proposal.md`, `design.md`, `specs/`, `tests.md`, `tasks.md`) — they exist locally and are tracked by `openspec status`. Only `tasks.md` is included in this PR diff because it's the only file that changed during implementation (task checkboxes were updated). The remaining artifacts don't change during the apply phase and don't need to be committed here. The archive step (post-merge) will commit the entire directory atomically.
